### PR TITLE
Replace gobig ec2 role

### PR DESCRIPTION
**Pull request by kotfic**
This replaces the gobig role with something that is less generic and more suited to cumulus' needs.  As an added bonus it should resolve https://github.com/Kitware/cumulus/issues/165 regardless of https://github.com/ansible/ansible-modules-core/issues/4011 by implementing an async based solution to terminating instances.  It adds a new integration test ```ec2_launch_cluster_integration_test.py``` that can be run like all other integration tests.  @cjh1 PTAL

### CODE REVIEW

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/defaults/main.yml [hunk 1]
```diff
@@ -1,26 +1,19 @@
...
 
-    default_public_key: ec2
-    default_image: ami-d05e75b8
-    default_instance_type: t2.medium
+# Very insecure!
+ec2_pod_rules:
+  - proto: all
+    from_port: all
+    to_port: all
+    cidr_ip: "{{ ssh_cidr_ip }}"
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/defaults/main.yml]: Can we not keep it locked down by default, only open port 22 for ssh?

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml [hunk 2]
```diff
@@ -0,0 +1,153 @@
+---
+# Note that we don't include master/node ami variables - these
+# could be conditionally checked, but because the next thing we
+# do is try to find the master/node ami if they aren't defined
+# it seems unessisary.
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: unessisary => unnecessary

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    count_tag:
+      ec2_pod: "{{ cluster_id }}"
+      ec2_pod_instance_name: head
+    # There can be only one...
+    exact_count: 1
+
+    # Misc
+    group: "{{ ec2_security_group }}"
+    placement_group: "{{ ec2_placement_group | default(omit) }}"
+    wait: yes
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: Why do we need to wait, don't we want the master and node instances to launch in parallel?

Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: Or is this just waiting for the placement group?

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/terminate.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    region: "{{ cluster_region }}"
+    zone: "{{ cluster_zone }}"
+    state: absent
+    wait: yes
+    wait_timeout: "{{ terminate_wait_timeout }}"
+  with_items: "{{ remote['instances'] | default([]) |map(attribute='id') | list }}"
+  async: "{{ terminate_wait_timeout }}"
+  poll: 0
+  register: terminate_jobs
+
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/terminate.yml]: terminate_jobs?

Comment (kotfic) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/terminate.yml]: Added note:
```
  # Because this is run async,  each return value is actually a 'job'
  # not the return value of the ec2 module.
```

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml (hunk 3 above)
Comment (kotfic) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: Nope this is a bug

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/defaults/main.yml (hunk 1 above)
Comment (kotfic) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/defaults/main.yml]: Actually the current logic is wrong,  I'm pushing a fix now - we should use the calculated defaults if ec2_pod_rules is not defined.

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      - proto: tcp # ssh
+        from_port: 22
+        to_port: 22
+        cidr_ip: "{{ ssh_cidr_ip }}"
+
+      - proto: all
+        group_name: "ec2_pod_{{cluster_id}}"
+        group_desc: "Ec2 security group for cluster {{ cluster_id }}"
+  when: ec2_pod_rules is not defined
+
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: So we are currently using 'extra_rules' of the form:
```json
[{
        'proto': 'tcp',
        'from_port': 7000,
        'to_port': 7000,
        'cidr_ip': source_ip
    }]

```

What will that change to when this change is merged?

Comment (kotfic) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: Yes,  currently this removes the extra_rules  variable - instead you should pass in your own custom ```ec2_pod_rules``` list that looks like:

```yaml
ec2_pod_rules:
  - proto: tcp # ssh
    from_port: 22
    to_port: 22
    cidr_ip: "{{ ssh_cidr_ip }}"

  - proto: "tcp"
    from_port: 7000
    to_port: 7000
    cidr_ip: "source_ip"

  - proto: all
    group_name: "ec2_pod_{{cluster_id}}"
    group_desc: "Ec2 security group for cluster {{ cluster_id }}"
```

Unless this is overly burdensome in which case I can add ```extra_vars``` back in - my feeling though is that even though this is a little more typing,  it is more explicit,  and prevents edge cases produced by merging rule sets.

**Comment by cjh1** [comment]
I am trying to run the ec2_launch_cluster_integration_test.py test and run into:

```
======================================================================
FAIL: test (__main__.AnsibleIntegrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "ec2_launch_cluster_integration_test.py", line 151, in test
    self.launch_cluster()
  File "ec2_launch_cluster_integration_test.py", line 125, in launch_cluster
    self._wait_for_status(status_url, 'running', timeout=300)
  File "ec2_launch_cluster_integration_test.py", line 72, in _wait_for_status
    'state is "%s"' % (status, r['status']))
AssertionError: Resource never moved into the "running" state, current state is "launching"
```

**Comment by cjh1** [comment]
I notice we are no longer including the cluster id in the instance names, which makes it hard to tell which instance belong to which clusters ( all the names are the same ).

**Comment by kotfic** [comment]
@cjh1 I have spent about an hour looking into this,  it is related to the ```wait_for``` module intermittently not completing. I am still digging

**Comment by kotfic** [comment]
@cjh1 PTAL

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml [hunk 6]
```diff
@@ -124,26 +124,34 @@
     instance_tags:
       Name: "ec2_pod_{{ cluster_id }}_node"
       ec2_pod: "{{ cluster_id }}"
-      ec2_pod_instance_name: node
+      ec2_pod_instance_name: data
     count_tag:
       ec2_pod: "{{ cluster_id }}"
-      ec2_pod_instance_name: node
+      ec2_pod_instance_name: data
     exact_count: "{{ node_instance_count }}"
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: Why node => data?

**Comment by cjh1** [comment]
Can we correct the spelling in the last commit message.

**Comment by cjh1** [comment]
Also it works! Awesome, I was able to run the test and it passed :tada: Thank you

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml (hunk 6 above)
Comment (kotfic) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: gridengine/site.yml expects the nodes to be in the data group. Group membership is determined (by the cumulus.py dynamic inventory) from the ec2_pod_instance_name tag.

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      set_fact:
+        master_instance_ami: "{{ master_amis.results[0]['ami_id'] }}"
+  when: master_instance_ami is not defined
+
+# Find the Node AMI if node_instance_ami is not specified
+- block:
+    - name: Find latest node AMI
+      ec2_ami_find:
+        state: available
+        sort: creationDaten
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: This typo is causing an issue in HPCCloud.

```
[16:20:01.910] ERROR: TASK: ec2-pod : Find latest node AMI [error]
{
  "msg": "value of sort must be one of: name,description,tag,architecture,block_device_mapping,creationDate,hypervisor,is_public,location,owner_id,platform,root_device_name,root_device_type,state,virtualization_type, got: creationDaten",
  "module_name": "ec2_ami_find",
  "changed": false,
  "_ansible_no_log": false,
  "failed": true
}

```

Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: We could modify the integration test to select an AMI using tags to catch this sort of error in the future.

File: cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml (hunk 5 above)
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: I am happy to update our configuration to pass this new configuration option. My only concern with this approach is it requires the user to know about certain internal ansible variables such as ssh_cidr_ip or the group_name.

Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/roles/ec2-pod/tasks/launch.yml]: After talking with @kotfic in person. We don't have to pass in ssh_cidr_ip, we can pass in an actual address for example 0.0.0.0/0 to open ssh to everyone. In the case of group_name we should allow this to be configured so that it can be used for this sort of configuration, rather than generate it. Otherwise its a little confusing for an application to have to use a group name for configuration which they didn't provide.
